apps/OboeTester/app/src/main/java/com/mobileer/audio_device/CommunicationDeviceSpinner.java

```diff
@@ -0,0 +1,126 @@
+
+import com.mobileer.oboetester.R;
+
+import java.util.List;
+
+public class CommunicationDeviceSpinner extends Spinner {
+    private static final int CLEAR_DEVICE_ID = 0;
+    private static final String TAG = CommunicationDeviceSpinner.class.getName();
+    private AudioDeviceAdapter mDeviceAdapter;
+    private AudioManager mAudioManager;
+    private Context mContext;
+    AudioDeviceInfo[] mCommDeviceArray = null;
+
+    public CommunicationDeviceSpinner(Context context){
+        super(context);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, int mode){
+        super(context, mode);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs){
+        super(context, attrs);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr){
+        super(context, attrs, defStyleAttr);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr, int mode){
+        super(context, attrs, defStyleAttr, mode);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr,
+                                      int defStyleRes, int mode){
+        super(context, attrs, defStyleAttr, defStyleRes, mode);
+        setup(context);
+    }
+
+    public CommunicationDeviceSpinner(Context context, AttributeSet attrs, int defStyleAttr,
+                                      int defStyleRes, int mode, Theme popupTheme){
+        super(context, attrs, defStyleAttr, defStyleRes, mode, popupTheme);
+        setup(context);
+    }
+
+    public AudioDeviceInfo[] getCommunicationsDevices() {
+        return mCommDeviceArray;
+    }
+
+    private void setup(Context context){
+        mContext = context;
+
+        mAudioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+
+        mDeviceAdapter = new AudioDeviceAdapter(context);
+        setAdapter(mDeviceAdapter);
+
+        // Add a default entry to the list and select it
+        mDeviceAdapter.add(new AudioDeviceListEntry(CLEAR_DEVICE_ID,
+                mContext.getString(R.string.auto_select)));
+        setSelection(0);
+        setupCommunicationDeviceListener();
+    }
+
+    @TargetApi(31)
+    private void setupCommunicationDeviceListener(){
+        // Note that we will immediately receive a call to onDevicesAdded with the list of
+        // devices which are currently connected.
+        mAudioManager.registerAudioDeviceCallback(new AudioDeviceCallback() {
+            @Override
+            public void onAudioDevicesAdded(AudioDeviceInfo[] addedDevices) {
+                updateDeviceList();
+            }
+
+            public void onAudioDevicesRemoved(AudioDeviceInfo[] removedDevices) {
+                updateDeviceList();
+                setSelection(0);
```

Comment:
updateDeviceList already this to the 0th device. Should we do this again?

Reply:
Thanks. I removed it.